### PR TITLE
fix: Make DOCKER_CONFIG available to buildctl

### PR DESCRIPTION
**Pull request by Kern--**
Issue #, if available:

N/A

*Description of changes:*

When povisioning a finch VM, the CLI inserts a line into the user's bash rc that points DOCKER_CONFIG to the user's finch dir. Finch-daemon running through systemd doesn't load the bash rc and so it doesn't get this config. As a result, build requests that go through finch-daemon don't pass this variable to buildctl so buildctl/buildkit is unable to build on base images in private repositories.

This change temporarily fixes this by inserting DOCKER_CONFIG into the finch-daemon process based on the owner of the socket. That way, buildctl is able to load the finch config and retrieve credentials.

This should eventually be replaced by somehow passing the credentials that come in to the build API directly to buildctl rather than loading credentials based on the config.

*Testing done:*

Verified the new unit tests.

Manually built an image that required credentials to see that it works after the change, but not before.

- [x] I've reviewed the guidance in CONTRIBUTING.md

#### License Acceptance

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by henry118 — APPROVED** [review]
LGTM as a short-term workaround. 

Long term we have to convert the creds from docker API to something that the BuildKit can take as input.

**Comment by pendo324** [comment]
Instead of using `os.Setenv` in the finch-daemon code itself, is it possible to add an `Environment=DOCKER_CONFIG=<path/to/.finch` statement to the service definition?

File: cmd/finch-daemon/main.go
Comment (swagatbora90) [inline cmd/finch-daemon/main.go:251]: Considering non-VM cases, should this check first if the DOCKER_CONFIG value is already set?

**Comment by Kern--** [comment]
@pendo324 

>Instead of using os.Setenv in the finch-daemon code itself, is it possible to add an Environment=DOCKER_CONFIG=<path/to/.finch> statement to the service definition?

Because with the proposed change, we don't know `<path/to/.finch>` statically. 

We have a choice of looking for config at one of:
1) MacOS/Windows/Linux specific finch dir (mounted from the host os in the MacOS/Windows cases)
2) The non-root user's $HOME/.finch (that has been symlinked to the host OS specific dir)
3) /root/.finch (that has been symlinked to the host OS specific dir)

Option 1 seems like even more complexity than the current changes.
Option 2 is what I chose here so that all finch related config ends up in the VM in the non-root user's home directory
Option 3 is doable, but it would mean finch CLI uses the user's home, but finch daemon uses root home even though the socket is owned by non root.

What do you think?



A side note: I wanted to use a systemd environment variable via `%h/.finch` but since we start the service as root, `%h` resolves to `/root`/.

**Comment by pendo324** [comment]
> @pendo324
> 
> > Instead of using os.Setenv in the finch-daemon code itself, is it possible to add an Environment=DOCKER_CONFIG=<path/to/.finch> statement to the service definition?
> 
> Because with the proposed change, we don't know `<path/to/.finch>` statically.

For all of the package deployments we currently own:
1. Amazon Linux (the "native" case) 
2. cases where Finch uses a VM (macOS/Windows)

We currently only support installations where there's a single user. In these scenarios, we can hardcode the DOCKER_CONFIG path in the systemd unit file, either as `%h/.finch` for Amazon Linux, or dynamically, via injecting the line in a Lima [provisioning script](https://github.com/runfinch/finch/pull/1181/files#diff-99291c33fb06ee1b2028a9f43494f71c272ea48c62052c929dea5cf6a33d8e0dR12-R14) (link is from my in-progress PR to add finch-daemon to Finch)

I'm not against merging this code, I'm just wondering if its worth adding it compared to the alternatives I just described

**Comment by Kern--** [comment]
No, I think if we only have a single user, it's probably better to hardcode the environment variable. That would also allow us to skip the symlink to make the finch directory at a known location inside the VM.

Should we cancel this and wait until your changes are merge or add this directly to your changes?

**Review by pendo324 — APPROVED** [review]

**Comment by pendo324** [comment]
> No, I think if we only have a single user, it's probably better to hardcode the environment variable. That would also allow us to skip the symlink to make the finch directory at a known location inside the VM.
> 
> Should we cancel this and wait until your changes are merge or add this directly to your changes?

I'm good to merge this as my changes will take longer to integrate across all of our deployments of finch-daemon, and we can always remove this functionality later if needed since finch-daemon is still pre-1.0.

File: cmd/finch-daemon/main.go
Comment (swagatbora90) [reply cmd/finch-daemon/main.go:259]: This is not a big concern, so can be added as a follow up. And may be irrelevant if we decide to hardcode the environment variable.

**Pull request merged**
